### PR TITLE
Feature Pyramid Grids

### PR DESCRIPTION
**Pull request by yhcao6**
Implementation of FPG that published in https://arxiv.org/abs/2004.03580.

Here is the mAP comparison of this PR and paper:
| Model              | Paper | This PR |
|-----------------------------------|-------|-------|
| Faster_R50_FPG_640                     | 41.4    | 42.2  |
| Faster_R50_FPG-cha128_640       | 40.0    | 41.2  |
| Mask_R50_FPG_640                      |  42.4   | 42.7  |
| Mask_R50_FPG-cha128_640        |   40.9  | 41.7   |
| Retinanet_R50_FPG_640               | 40.0   | 40.5   |
| Retinanet_R50_FPG-cha128_640 |  39.0  |   |

### CODE REVIEW

File: mmdet/models/necks/fpg.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+        self.out_channels = out_channels
+
+    def forward(x):
+        pass
+
+
+class UpInterpolationConv(Transition):
+    """A transition used for up-sampling.
+
+    Firstly up-sampling the input by interpolation and follow by a
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:28]: Up-sample the input by interpolation then refines the feature by a convolution layer.

File: mmdet/models/necks/fpg.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    conv for refining.
+
+    Args:
+        in_channels (int): Number of input channels.
+        out_channels (int): Number of output channels.
+        scale_factor (int): Up-sampling factor. Default: 2.
+        mode (int): Interpolation mode. Default: nearest.
+        align_corners (bool): Whether align corners when interpolation.
+            Default: None.
+        kernel_size (int): Kernel size for the conv.
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:38]: Default: 3.

File: mmdet/models/necks/fpg.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 scale_factor=2,
+                 mode='nearest',
+                 align_corners=None,
+                 kernel_size=3,
+                 **kwargs):
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:48]: Can we use `conv_cfg` here?

File: mmdet/models/necks/fpg.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+
+class LastConv(Transition):
+    """A transition used for refining the output of the last stage.
+
+    Args:
+        in_channels (int): Number of input channels.
+        out_channels (int): Number of output channels.
+        num_inputs (int): Number of inputs of the FPN features.
+        kernel_size (int): Kernel size for the conv.
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:77]: Default

File: mmdet/models/necks/fpg.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+
+
+@NECKS.register_module()
+class FPG(nn.Module):
+    """FPG.
+
+    Implementation of `Feature Pyramid Grids (FPG)
+    <https://arxiv.org/abs/2004.03580>`_.
+    This implementation only gives the basic structure stated in the paper.
+    But users can implement different transition type to fully explore the
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:107]: different type of transitions

**Review by ZwwWayne — APPROVED** [review]

File: mmdet/models/necks/fpg.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+            build the feature pyramid. Default: 0.
+        end_level (int): Index of the end input backbone level (exclusive) to
+            build the feature pyramid. Default: -1, which means the last level.
+        add_extra_convs (bool): It decides whether to add conv
+            layers on top of the original feature maps. Default to False.
+            If True, its actual mode is specified by `extra_convs_on_inputs`.
+        norm_cfg (dict): Config dict for normalization layer. Default: None.
+    """
+
+    transition_types = {
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:138]: Can we create a new registry for transition thus we could extend the transitions without modifying the code in fpg.py?

File: mmdet/models/necks/fpg.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+        cfg_ = cfg.copy()
+        trans_type = cfg_.pop('type')
+        trans_cls = self.transition_types[trans_type]
+        return trans_cls(in_channels, out_channels, **cfg_, **extra_args)
+
+    def init_weights(self):
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                caffe2_xavier_init(m)
+            elif isinstance(m,
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:311]: May directly use `is_norm` in mmcv.cnn [here](https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/norm.py#L121).

File: mmdet/models/necks/fpg.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+            current_outs = outs[-1]
+            next_outs = []
+            direction = self.paths[i]
+            for j in range(self.num_outs):
+                if i in self.skip_inds[j]:
+                    next_outs.append(outs[-1][j])
+                    continue
+                # feature level
+                if direction == 'td':
+                    lvl = self.num_outs - j - 1
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:348]: May add comments indicating what `lvl` stands for.

File: mmdet/models/necks/fpg.py [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+                    same_trans = self.fpn_transitions[i][lvl]['same_down']
+                else:
+                    same_trans = self.fpn_transitions[i][lvl]['same_up']
+                across_lateral_trans = self.fpn_transitions[i][lvl][
+                    'across_lateral']
+                across_down_trans = self.fpn_transitions[i][lvl]['across_down']
+                across_up_trans = self.fpn_transitions[i][lvl]['across_up']
+                across_skip_trans = self.fpn_transitions[i][lvl]['across_skip']
+                # init output
+                # x = torch.zeros_like(current_outs[lvl])
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:362]: may clean code.

File: mmdet/models/necks/fpg.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+                 add_extra_convs=False,
+                 norm_cfg=None,
+                 skip_inds=None):
+        super(FPG, self).__init__()
+        assert isinstance(in_channels, list)
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.num_ins = len(in_channels)
+        self.num_outs = num_outs
+        if inter_channels is None:
```
Comment (ZwwWayne) [inline mmdet/models/necks/fpg.py:170]: Do we need a unit test to more comprehensively test this module?

File: configs/fpg/README.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+## Results and Models
+
+We benchmark the new training schedule (crop training, large batch, unfrozen BN, 50 epochs) introduced in NAS-FPN.
+All backbones are Resnet-50 in pytorch style.
+
+| Method       | Neck        | Lr schd | Mem (GB) | Inf time (fps) | box AP | mask AP | Config | Download |
+|:------------:|:-----------:|:-------:|:--------:|:--------------:|:------:|:-------:|:-------:|:--------:|
+| Faster R-CNN | FPG         | 50e     | 13.2     | -              | 42.2   | -       |[config]() |
+| Faster R-CNN | FPG-cha128  | 50e     | 13.2     | -              | 41.2   | -       |[config]() |
```
Comment (hellock) [inline configs/fpg/README.md:22]: - We may use `chn` instead of `cha`.
- Config file path is empty.

File: mmdet/models/necks/fpg.py [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+    the potential power of the structure of FPG.
+
+    Args:
+        in_channels (int): Number of input channels (feature maps of all levels
+            should have the same channels).
+        out_channels (int): Number of output channels (used at each scale)
+        num_outs (int): Number of output scales.
+        stack_times (int): The number of times the pyramid architecture will
+            be stacked.
+        paths (list(str)): Specify the top-down or bottom-up order of each
```
Comment (hellock) [inline mmdet/models/necks/fpg.py:117]: - list[str]
- Specify the path order of each stack level. Each element in the list should be either 'bu' (bottom-up) or 'td' (top-down).

File: mmdet/models/necks/fpg.py (hunk 8 above)
Comment (yhcao6) [reply mmdet/models/necks/fpg.py:348]: The comment is there `# feature level`

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**